### PR TITLE
4 camera

### LINKED ISSUE #4: Camera
The camera defines the image of the scene/world that is perceived. Therefore all factors that pertain to the image description are handled by the camera.

Camera factors:

- aspect ratio (image width/image height)
- image width -> height derived with ratio
- viewport dimensions
- camera position (from file)
- viewing direction (from file)
- view up vector -> derived from viewing direction
- field of view (from file)
- ray generation and collection ? (transfer to renderer?)

### PR DESCRIPTION
**Pull request by u413-284-si**
First static camera implemented with

- image size and ratio definition
- viewport position and parameters
- image creation by sequentially passing rays through each pixel

First scene modeled with
- 2 spheres
- colour assignment dependent on normal vector direction
- background colouring

Hit management of spheres
- hit record
- optimisation on rendering closest hit

Vector manipulation functions

Utility functions

closes #4

### CODE REVIEW

File: inc/camera.h [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+ * 
+ * The viewport is a virtual rectangle in the 3D world containing the grid
+ * of image pixel locations. If pixels are equispaced horizontally and verti-
+ * cally, the viewport has the same aspect_ratio as the image.
+ * The focal length is the distance between the camera centre and the viewport
+ * centre (vector orthogonal on viewport plane).
+ * @param aspect_ratio			Ratio of image_width / image_height
+ * @param image_width			Width of image created
+ * @param image_height			Height of image created
+ * @param viewport_height		Height of viewport rectangle
```
Comment (gwolf-011235) [inline inc/camera.h]: Nit: keep order the same -> width then height or height then width

File: inc/hit_manager.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ * @param rec		Hit record of ray with sphere
+ * @param ray_d 	Interval of accepted values of the ray distance for a hit
+ * @return true		If sphere is hit
+ * @return false	If sphere is missed
+ */
+bool	ft_hit_sphere(t_sphere sp, t_ray ray, t_hitrecord *rec, \
+	t_interval ray_d);
+
+/**
+ * @brief
```
Comment (gwolf-011235) [inline inc/hit_manager.h]: Missing description

File: inc/ray.h [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+{
+	t_vec3	point;
+	t_vec3	normal;
+	float	d;
+}	t_hitrecord;
+
+/* ====== DECLARATIONS ====== */
+
+/**
+ * @brief
```
Comment (gwolf-011235) [inline inc/ray.h]: Missing descriptions :heart:

File: inc/utilities.h [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+/*   utilities.h                                        :+:      :+:    :+:   */
+/*                                                    +:+ +:+         +:+     */
+/*   By: u413q <u413q@student.42.fr>                +#+  +:+       +#+        */
+/*                                                +#+#+#+#+#+   +#+           */
+/*   Created: 2023/10/29 11:46:04 by u413q             #+#    #+#             */
+/*   Updated: 2023/10/29 17:56:01 by u413q            ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+#ifndef UTILITIES_H
```
Comment (gwolf-011235) [inline inc/utilities.h]: Nit: i think "utils" is more common

File: src/utilities.c [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+/*   Created: 2023/10/29 11:48:18 by u413q             #+#    #+#             */
+/*   Updated: 2023/10/29 17:56:09 by u413q            ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "utilities.h"
+
+float	ft_degree_to_radian(float degrees)
+{
+	return (degrees * M_PI / 180.0);
```
Comment (gwolf-011235) [inline src/utilities.c]: Performance improvement: define a conversion constant to multiply with.

E.G: # define RAD 0.01745329251

File: src/vec3_arithmetics.c [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+/*   Created: 2023/10/26 13:56:24 by u413q             #+#    #+#             */
+/*   Updated: 2023/10/26 14:52:36 by u413q            ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "vec3.h"
+
+t_vec3	ft_vec3_add(t_vec3 v1, t_vec3 v2)
+{
+	t_vec3	rtrn;
```
Comment (gwolf-011235) [inline src/vec3_arithmetics.c:17]: Nit: ret instead of rtrn.

**Review by gwolf-011235 — APPROVED** [review]
Looks very good!
Im impressed with the rendered sphere.
Only found some Nits and style stuff
Noice

File: inc/camera.h (hunk 1 above)
Comment (u413-284-si) [reply inc/camera.h]: Changed

File: inc/hit_manager.h (hunk 2 above)
Comment (u413-284-si) [reply inc/hit_manager.h]: added

File: inc/ray.h (hunk 3 above)
Comment (u413-284-si) [reply inc/ray.h]: added

File: inc/utilities.h (hunk 4 above)
Comment (u413-284-si) [reply inc/utilities.h]: adapted in each occurence and file names .c/.h

File: src/utilities.c (hunk 5 above)
Comment (u413-284-si) [reply src/utilities.c]: added

File: src/vec3_arithmetics.c (hunk 6 above)
Comment (u413-284-si) [reply src/vec3_arithmetics.c:17]: denied. it's buuutiful

Comment (gwolf-011235) [reply src/vec3_arithmetics.c:17]: Nooooooo. ok

**Pull request merged**